server/resource_server/service/service.go

```diff
@@ -172,7 +179,54 @@ func (s *service) ProfileClient(userName string) (models.ClientResponseOutput, e
 }
 
 func (s *service) VerifyEmail(userID uint) error {
-	return s.repository.VerifyEmail(userID)
+	user, e := s.repository.FindUser(userID)
+	if e != nil {
+		return e
+	}
+
+	verificationCode := s.emailVerifier.GenerateVerificationCode(&user)
+
+	e = s.emailVerifier.SendVerificationEmail(&user, verificationCode)
+	if e != nil {
+		return e
+	}
+
+	e = s.repository.SaveEmailVerificationData(
+		models.EmailVerificationData{
+			UserId:           user.ID,
+			VerificationCode: verificationCode,
+			ExpiresAt:        time.Now().Add(constants.VerificationCodeValidityDuration),
+		},
+	)
+
+	return e
+}
+
+func (s *service) VerifyCode(userId uint, code string) error {
```

Comment:
a lots of interaction with db is happens here, we definitely need to wrap this in a transaction

Reply:
Transaction is generally a good thing to do, although in this case I don't think it's mandatory. Gorm.db executes all the modification queries within a transaction by itself, so there wouldn't be any problems with other parallel queries interacting with the same tables. The only issue might have been if this function is triggered twice for the same user (in the current flow this might happen if the user clicks the Submit button twice), but in this case there would be no inconsistencies because the data we save (i.e. verification codes and zk-proofs) is unique per user and thus no overwrites would occur.
But I agree that we might actually need to add it in the future, depending on how the flow changes.

Reply:
the purpose of transaction here is not for locking but for atomicity purpose (https://www.beekeeperstudio.io/blog/what-is-atomicity-in-databases). As an example if somehow the user fail to change the verification status due to network or intermittent issue, then it will leads to data incosistency because we already delete the verification data from the db

Reply:
@interestIngc I'll let you decide between "good" and "good enough" here.